From 8799eaa56047b3b56c9e2e87b002189eda8faffa Mon Sep 17 00:00:00 2001
From: Kiran Kelageri <kirankelageri@codeaurora.org>
Date: Mon, 15 Apr 2019 15:28:33 -0700
Subject: [PATCH] Bluetooth-SWB: Power managment based SWB.

Handle SWB codec selection based on network type
and codecs selected over network during call
initiation.

CRs-fixed: 2443069
Change-Id: I9bc4bb27e36b4592405b275dad64c8f8e3a00857
---
 .../android/bluetooth/hfp/HeadsetService.java | 28 ++++++++++++++++
 .../bluetooth/hfp/HeadsetStateMachine.java    | 32 +++++++++++++++++++
 .../bluetooth/hfp/HeadsetSystemInterface.java | 17 ++++++++++
 3 files changed, 77 insertions(+)

diff --git a/src/com/android/bluetooth/hfp/HeadsetService.java b/src/com/android/bluetooth/hfp/HeadsetService.java
index f879d56bb..7ccabf41d 100644
--- a/src/com/android/bluetooth/hfp/HeadsetService.java
+++ b/src/com/android/bluetooth/hfp/HeadsetService.java
@@ -128,6 +128,8 @@ public class HeadsetService extends ProfileService {
     private boolean mIsTwsPlusEnabled = false;
     private boolean mIsTwsPlusShoEnabled = false;
     private vendorhfservice  mVendorHf;
+    private boolean mIsSwbEnabled = false;
+    private boolean mIsSwbPmEnabled = false;
 
     @Override
     public IProfileServiceBinder initBinder() {
@@ -166,6 +168,8 @@ public class HeadsetService extends ProfileService {
             String twsPlusEnabled = SystemProperties.get("persist.vendor.btstack.enable.twsplus");
             String twsPlusShoEnabled =
                SystemProperties.get("persist.vendor.btstack.enable.twsplussho");
+            String swbEnabled = SystemProperties.get("persist.vendor.btstack.enable.swb");
+            String swbPmEnabled = SystemProperties.get("persist.vendor.btstack.enable.swbpm");
             if (!twsPlusEnabled.isEmpty() && "true".equals(twsPlusEnabled)) {
                 mIsTwsPlusEnabled = true;
             }
@@ -177,8 +181,16 @@ public class HeadsetService extends ProfileService {
                     mIsTwsPlusShoEnabled = false;
                 }
             }
+            if (!swbEnabled.isEmpty() && "true".equals(swbEnabled)) {
+                mIsSwbEnabled = true;
+            }
+            if (!swbPmEnabled.isEmpty() && "true".equals(swbPmEnabled)) {
+                mIsSwbPmEnabled = true;
+            }
             Log.i(TAG, "mIsTwsPlusEnabled: " + mIsTwsPlusEnabled);
             Log.i(TAG, "mIsTwsPlusShoEnabled: " + mIsTwsPlusShoEnabled);
+            Log.i(TAG, "mIsSwbEnabled: " + mIsSwbEnabled);
+            Log.i(TAG, "mIsSwbPmEnabled: " + mIsSwbPmEnabled);
             if (mIsTwsPlusEnabled && mMaxHeadsetConnections < 2){
                //set MaxConn to 2 if TWSPLUS enabled
                mMaxHeadsetConnections = 2;
@@ -222,7 +234,9 @@ public class HeadsetService extends ProfileService {
         mHfpA2dpSyncInterface = new HeadsetA2dpSync(mSystemInterface, this);
         if (mVendorHf != null) {
             mVendorHf.init();
+            mVendorHf.enableSwb(mIsSwbEnabled);
         }
+
         Log.i(TAG, " HeadsetService Started ");
         return true;
     }
@@ -1223,6 +1237,7 @@ public class HeadsetService extends ProfileService {
                 stateMachine.sendMessage(HeadsetStateMachine.VOICE_RECOGNITION_START, device);
             }
         }
+        enableSwbCodec(true);
         return true;
     }
 
@@ -1864,6 +1879,7 @@ public class HeadsetService extends ProfileService {
             if (!mSystemInterface.getVoiceRecognitionWakeLock().isHeld()) {
                 mSystemInterface.getVoiceRecognitionWakeLock().acquire(sStartVrTimeoutMs);
             }
+            enableSwbCodec(true);
             return true;
         }
     }
@@ -2407,6 +2423,18 @@ public class HeadsetService extends ProfileService {
         }
     }
 
+    public void enableSwbCodec(boolean enable) {
+        mVendorHf.enableSwb(enable);
+    }
+
+    public boolean isSwbEnabled() {
+        return mIsSwbEnabled;
+    }
+
+    public boolean isSwbPmEnabled() {
+        return mIsSwbPmEnabled;
+    }
+
     private static void logD(String message) {
         if (DBG) {
             Log.d(TAG, message);
diff --git a/src/com/android/bluetooth/hfp/HeadsetStateMachine.java b/src/com/android/bluetooth/hfp/HeadsetStateMachine.java
index 5258c2efa..25750c9be 100644
--- a/src/com/android/bluetooth/hfp/HeadsetStateMachine.java
+++ b/src/com/android/bluetooth/hfp/HeadsetStateMachine.java
@@ -1399,6 +1399,18 @@ public class HeadsetStateMachine extends StateMachine {
                                   + " or A2Dp is playing, not allowing SCO, device=" + mDevice);
                         break;
                     }
+
+                    if (mHeadsetService.isSwbEnabled() && mHeadsetService.isSwbPmEnabled()) {
+                        if (!mHeadsetService.isVirtualCallStarted() &&
+                             mSystemInterface.isHighDefCallInProgress()) {
+                           log("CONNECT_AUDIO: enable SWB for HD call ");
+                           mHeadsetService.enableSwbCodec(true);
+                        } else {
+                           log("CONNECT_AUDIO: disable SWB for non-HD or Voip calls");
+                           mHeadsetService.enableSwbCodec(false);
+                        }
+                    }
+
                     if (!mNativeInterface.connectAudio(mDevice)) {
                         stateLogE("Failed to connect SCO audio for " + mDevice);
                         // No state change involved, fire broadcast immediately
@@ -2183,6 +2195,22 @@ public class HeadsetStateMachine extends StateMachine {
                 + callState.mNumHeld + " mCallState: " + callState.mCallState);
         log("processCallState: mNumber: " + callState.mNumber + " mType: " + callState.mType);
 
+        if (mHeadsetService.isSwbEnabled() && mHeadsetService.isSwbPmEnabled()) {
+            if (mHeadsetService.isVirtualCallStarted()) {
+                 log("processCallState: enable SWB for all voip calls ");
+                 mHeadsetService.enableSwbCodec(true);
+            } else if((callState.mCallState == HeadsetHalConstants.CALL_STATE_DIALING) ||
+               (callState.mCallState == HeadsetHalConstants.CALL_STATE_INCOMING)) {
+                 if (!mSystemInterface.isHighDefCallInProgress()) {
+                    log("processCallState: disable SWB for non-HD call ");
+                    mHeadsetService.enableSwbCodec(false);
+                 } else {
+                    log("processCallState: enable SWB for HD call ");
+                    mHeadsetService.enableSwbCodec(true);
+                 }
+            }
+        }
+
         processA2dpState(callState);
     }
 
@@ -2299,6 +2327,9 @@ public class HeadsetStateMachine extends StateMachine {
         switch (wbsConfig) {
             case HeadsetHalConstants.BTHF_WBS_YES:
                 mAudioParams.put(HEADSET_WBS, HEADSET_AUDIO_FEATURE_ON);
+                if (mHeadsetService.isSwbEnabled()) {
+                    mAudioParams.put(HEADSET_SWB, HEADSET_SWB_DISABLE);
+                }
                 break;
             case HeadsetHalConstants.BTHF_WBS_NO:
             case HeadsetHalConstants.BTHF_WBS_NONE:
@@ -2315,6 +2346,7 @@ public class HeadsetStateMachine extends StateMachine {
     private void processSWBEvent(int swbConfig) {
         if (swbConfig < HEADSET_SWB_MAX_CODEC_IDS) {
                 mAudioParams.put(HEADSET_SWB, "0");
+                mAudioParams.put(HEADSET_WBS, HEADSET_AUDIO_FEATURE_OFF);
         } else {
                 mAudioParams.put(HEADSET_SWB, HEADSET_SWB_DISABLE);
         }
diff --git a/src/com/android/bluetooth/hfp/HeadsetSystemInterface.java b/src/com/android/bluetooth/hfp/HeadsetSystemInterface.java
index 16ea63516..e4078ac59 100644
--- a/src/com/android/bluetooth/hfp/HeadsetSystemInterface.java
+++ b/src/com/android/bluetooth/hfp/HeadsetSystemInterface.java
@@ -237,6 +237,23 @@ public class HeadsetSystemInterface {
         return false;
     }
 
+    /**
+     * Check for HD codec for voice call
+     */
+    @VisibleForTesting
+    public boolean isHighDefCallInProgress() {
+        if (mPhoneProxy != null) {
+            try {
+                return mPhoneProxy.isHighDefCallInProgress();
+            } catch (RemoteException e) {
+                Log.e(TAG, Log.getStackTraceString(new Throwable()));
+            }
+        } else {
+            Log.e(TAG, "Handsfree phone proxy null");
+        }
+        return false;
+    }
+
     /**
      * Get the the alphabetic name of current registered operator.
      *
-- 
2.17.1

